gammapy/background/reflected.py

```diff
@@ -135,60 +86,92 @@ def __init__(
         """
-        binsz = Angle(binsz)
-        width = 3.0 * region.center.separation(center)
+        width = 15.
+
+        try:
+            reg_center = self.region.center
+        except:
+            raise NotImplementedError("Algorithm not yet adapted to this Region shape")
+
+        if 'ra' in reg_center.representation_component_names:
+            _maskmap = WcsNDMap.create(
+                skydir=self.center, binsz=self.binsz, width=width, coordsys="CEL", proj="TAN"
+             )
+        else:
+            _maskmap = WcsNDMap.create(
+                skydir=self.center, binsz=self.binsz, width=width, coordsys="GAL", proj="TAN"
+            )
+
+        # width is the full width of an image (not the radius)
+        width = Angle(3.0 * reg_center.transform_to(self.center).separation(self.center), u.degree) * 2.
+        maskmap = _maskmap.cutout(self.center, width)
 
-        maskmap = WcsNDMap.create(
-            skydir=center, binsz=binsz, width=width, coordsys="GAL", proj="TAN"
-        )
         return maskmap
 
     def setup(self):
         """Compute parameters for reflected regions algorithm."""
-        wcs = self.reference_map.geom.wcs
-        self._pix_region = self.region.to_pixel(wcs)
-        self._pix_center = PixCoord(*self.center.to_pixel(wcs))
+        geom = self.reference_map.geom
+        self._pix_region = self.region.to_pixel(geom.wcs)
+        self._pix_center = PixCoord.from_sky(self.center, geom.wcs)
         dx = self._pix_region.center.x - self._pix_center.x
         dy = self._pix_region.center.y - self._pix_center.y
 
+        if self._pix_region.contains(self._pix_center):
+            log.warn("Obs #{} rejected! Pointing position within the ON region".format(self.obs_id))
+            return
+
         # Offset of region in pix coordinates
         self._offset = np.hypot(dx, dy)
 
+        # Make the ON reference map
+        _mask = geom.region_mask([self.region], inside=True)
+        self.on_reference_map = WcsNDMap(geom=geom, data=_mask)
+        # TODO: remove this lign after the correction of the Issue #2074
+        self.on_reference_map.data = self.on_reference_map.data.astype(float)
+
         # Starting angle of region
-        self._angle = Angle(np.arctan2(dx, dy), "rad")
+        self._angle = Angle(np.arctan2(dy, dx), "rad")
 
         # Minimum angle a circle has to be moved to not overlap with previous one
-        min_ang = Angle(2 * np.arcsin(self._pix_region.radius / self._offset), "rad")
+        pix_idx = geom.get_pix()
+        pix_on_x = pix_idx[0][_mask]
+        pix_on_y = pix_idx[1][_mask]
+        newX, newY = (self._pix_center.x - pix_on_x), (self._pix_center.y - pix_on_y)
+        angles = np.arctan2(newY, newX) * u.rad
+        min_ang = np.max(angles)-np.min(angles)
+        if min_ang.value > np.pi:
+            new_angles = np.zeros_like(angles)
+            new_angles[angles.value > 0] = angles[angles.value > 0] - np.pi*u.rad
```

Comment:
Can this be simplified? `astropy.coordinates.Angle` has methods that implement such wrapping logic?

Reply:
Except using the class Longitude, I do not see how to to... Should I use this class?

Reply:
Try to use the `Angle.wrap_at` method.
```
>>> from astropy.coordinates import Angle
>>> a = Angle([-200, -300, 0, 100, 200, 300, 400], unit='deg')
>>> a.wrap_at('180 deg')
<Angle [ 160.,   60.,    0.,  100., -160.,  -60.,   40.] deg>
```

Reply:
done